### PR TITLE
Update cluster upgrade

### PR DESCRIPTION
**Pull request by flamingdumpster**
**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ x ] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
Currently , the pgo upgrade command only updates the primary and requires replicas to be updated manually.

**What is the new behavior (if this is a feature change)?**
- pgo upgrade updates the replicas and the primary, in a sequential order
- replicas are updated first one at a time
- primary is updated last after all replicas
- presently, backrest-repo container is part of operator release, not updated by pgo upgrade.

**Other information**:
[ch5448]

### CODE REVIEW

**Comment by flamingdumpster** [comment]
@tjmoore4 unless _AutofailReplaceReplica_ is set to true in the pgo.yaml, replicas will not be replaced after a failover. I believe the behavior you noted is expected as that value defaults to false.

**Comment by tjmoore4** [comment]
@flamingdumpster retried with `auto_failover_replace_replica=true` and it worked as expected, thanks.

File: operator/cluster/upgrade.go [hunk 1]
```diff
@@ -16,52 +16,366 @@ package cluster
...
+			return err
+		}
+
+		return err
+}
+
+
+func removeMinorUpgradeLabelFromCluster(clientset *kubernetes.Clientset, restclient *rest.RESTClient, clusterName, namespace string) error {
+
+	// TODO - when upgrade completes, this gets called to remove the label.
```
Comment (jkatz) [inline operator/cluster/upgrade.go]: Is this TODO still relevant?

File: postgres-operator.go [hunk 2]
```diff
@@ -181,7 +181,7 @@ func main() {
 
 	cluster.InitializeAutoFailover(Clientset, crdClient, namespaceList)
 
-	operatorupgrade.OperatorUpgrade(Clientset, crdClient, namespaceList)
+	operatorupgrade.OperatorUpdateCRPgoVersion(Clientset, crdClient, namespaceList)
```
Comment (jkatz) [inline postgres-operator.go]: Is there a reason for the function name change? Not sure I follow what This is doing.

File: apiserver/upgradeservice/upgradeimpl.go [hunk 3]
```diff
@@ -24,6 +24,8 @@ import (
 	log "github.com/sirupsen/logrus"
 	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	// "github.com/crunchydata/postgres-operator/util"
```
[the remark this answers is not in the document]
Comment (flamingdumpster) [reply apiserver/upgradeservice/upgradeimpl.go]: done

File: apiserver/upgradeservice/upgradeimpl.go [hunk 4]
```diff
@@ -113,7 +115,7 @@ func CreateUpgrade(request *msgs.CreateUpgradeRequest, ns string) msgs.CreateUpg
 			response.Status.Msg = clusterName + " is not a valid pgcluster"
 			return response
 		}
-
+
```
[the remark this answers is not in the document]
Comment (flamingdumpster) [reply apiserver/upgradeservice/upgradeimpl.go]: removed an extra line.

File: controller/podcontroller.go [hunk 5]
```diff
@@ -115,8 +115,18 @@ func (c *PodController) onUpdate(oldObj, newObj interface{}) {
...
+
+	// have a pod coming back up from upgrade and is ready - time to kick off the next pod.
+	if clusterInMinorUpgrade && isUpgradedPostgresPod(newpod, oldpod) {
+		upgradeTaskName := clusterName + "-" + config.LABEL_MINOR_UPGRADE
+		clusteroperator.ProcessNextUpgradeItem(c.PodClientset, c.PodClient, clusterName, upgradeTaskName, newpod.ObjectMeta.Namespace)
+	}
+
 	//handle the case when a pg database pod is updated
-	if isPostgresPod(newpod) {
+	if isPostgresPod(newpod) {
```
[the remark this answers is not in the document]
Comment (flamingdumpster) [reply controller/podcontroller.go]: corrected

File: controller/podcontroller.go [hunk 6]
```diff
@@ -293,31 +304,96 @@ func (c *PodController) checkAutofailLabel(newpod *apiv1.Pod, ns string) bool {
...
 		return false
 	}
 	if newpod.ObjectMeta.Labels[config.LABEL_PGBOUNCER] == "true" {
-		log.Debugf("pgbouncer pod found [%s]", newpod.Name)
+		// log.Debugf("pgbouncer pod found [%s]", newpod.Name)
 		return false
 	}
 	return true
 }
+// isUpgradedPostgresPod - determines of the pod is one that could be getting a minor upgrade
```
[the remark this answers is not in the document]
Comment (flamingdumpster) [reply controller/podcontroller.go]: corrected

File: controller/podcontroller.go [hunk 7]
```diff
@@ -293,31 +304,96 @@ func (c *PodController) checkAutofailLabel(newpod *apiv1.Pod, ns string) bool {
 
 func isPostgresPod(newpod *apiv1.Pod) bool {
 	if newpod.ObjectMeta.Labels[config.LABEL_PGO_BACKREST_REPO] == "true" {
-		log.Debugf("pgo-backrest-repo found %s", newpod.Name)
+	// log.Debugf("pgo-backrest-repo found %s", newpod.Name)
```
[the remark this answers is not in the document]
Comment (flamingdumpster) [reply controller/podcontroller.go]: left in.

File: operator/cluster/failoverlogic.go [hunk 8]
```diff
@@ -170,14 +170,14 @@ func deletePrimary(clientset *kubernetes.Clientset, namespace, clusterName, pgou
...
-	}
-	if len(pods.Items) > 1 {
-		log.Errorf("more than 1 primary pod found when trying to delete primary %s", selector)
-		return errors.New("more than 1 primary pod found in delete primary logic")
+	// pods, err := kubeapi.GetPods(clientset, selector, namespace)
+
+	pods, success := waitForDeploymentPrimaryCount(clientset, selector, namespace)
+
+
+	if !success {
```
[the remark this answers is not in the document]
Comment (flamingdumpster) [reply operator/cluster/failoverlogic.go]: It is poorly named also, will rename method and adjust comments accordingly.

File: controller/podcontroller.go [hunk 9]
```diff
@@ -293,31 +304,96 @@ func (c *PodController) checkAutofailLabel(newpod *apiv1.Pod, ns string) bool {
...
+			return true
+		}
+
+		if newpod.ObjectMeta.Labels[config.LABEL_SERVICE_NAME] == replicaServiceName {
+			log.Debugf("Minor Upgrade: upgraded replica found %s", newpod.Name)
+			return true
+		}
+
+		// This indicates there is a pod we didn't account for - shouldn't be the case
+		log.Debugf(" **** Minor Upgrade: unexpected isUpgraded pod found: [%s] ****", newpod.Name)
```
[the remark this answers is not in the document]
Comment (flamingdumpster) [reply controller/podcontroller.go]: No, its a more of "look at this when it happens, was there something we missed" type of logging. Probably means we didn't filter out something we should have.

File: operator/cluster/failoverlogic.go [hunk 10]
```diff
@@ -170,14 +170,14 @@ func deletePrimary(clientset *kubernetes.Clientset, namespace, clusterName, pgou
...
 	selector := config.LABEL_SERVICE_NAME + "=" + clusterName + "," + config.LABEL_BACKREST_RESTORE + "!=true," + config.LABEL_BACKREST_JOB + "!=true"
-	pods, err := kubeapi.GetPods(clientset, selector, namespace)
-	if len(pods.Items) == 0 {
-		log.Errorf("no primary pod found when trying to delete primary %s", selector)
-		return errors.New("could not find primary pod")
-	}
-	if len(pods.Items) > 1 {
-		log.Errorf("more than 1 primary pod found when trying to delete primary %s", selector)
-		return errors.New("more than 1 primary pod found in delete primary logic")
+	// pods, err := kubeapi.GetPods(clientset, selector, namespace)
```
[the remark this answers is not in the document]
Comment (flamingdumpster) [reply operator/cluster/failoverlogic.go]: done

File: operator/cluster/failoverlogic.go [hunk 11]
```diff
@@ -170,14 +170,14 @@ func deletePrimary(clientset *kubernetes.Clientset, namespace, clusterName, pgou
...
+
+	pods, success := waitForDeploymentPrimaryCount(clientset, selector, namespace)
```
[the remark this answers is not in the document]
Comment (flamingdumpster) [reply operator/cluster/failoverlogic.go]: Done

File: operator/cluster/upgrade.go [hunk 12]
```diff
@@ -16,52 +16,366 @@ package cluster
...
 
-// AddUpgrade bounces the Deployment with the new image tag
+// AddUpgrade implements the upgrade workflow for cluster minor upgrade
 func AddUpgrade(clientset *kubernetes.Clientset, restclient *rest.RESTClient, upgrade *crv1.Pgtask, namespace string) {
-	cl := crv1.Pgcluster{}
 
-	//not a db so get the pgcluster CRD
-	_, err := kubeapi.Getpgcluster(restclient, &cl,
-		upgrade.ObjectMeta.Labels[config.LABEL_PG_CLUSTER], namespace)
+	// TODO: persist original and target CCP_IMAGE_TAGs in pgtask - think so...
```
[the remark this answers is not in the document]
Comment (flamingdumpster) [reply operator/cluster/upgrade.go]: removed.

File: operator/cluster/upgrade.go (hunk 1 above)
Comment (flamingdumpster) [reply operator/cluster/upgrade.go]: removed.

File: postgres-operator.go (hunk 2 above)
Comment (flamingdumpster) [reply postgres-operator.go]: it expands to "Operator Update Custom Resource PGO Version." I changed it to make it more distinct than prior name (which generally referred to several different upgrades or update) - the function is commented at its definition.

File: hugo/content/Upgrade/_index.md [hunk 13]
```diff
@@ -17,3 +17,15 @@ This section of the documentation shows specific steps required to upgrade diffe
...
+
+The upgrade is accomplished by updating the CCPImageTag version in the deployment, which causes the old pod to be terminated and a new pod created with the updated deployment specification.
+
+When the upgrade starts, each replica is upgraded seqentially, waiting for the previous replica to go ready before updating the next. After the replicas complete, the primary is then upgraded to the new image. The upgrade process respects the _autofail_ and the _AutofailReplaceReplica_ settings as provided in the pgo.yaml or as a command line flag, if applicable.
+
+When the cluster is not in _autofail_ mode, the deployments simply create a new pod when updated, terminating the old one. When autofail is enabled and the primary deployment is updated, the cluster behaves as though the primary had failed and begins the failover process. See _Automatic Failover_ in the _Overview_ section for more details on this and replica replacement.
+
+At this time, the backrest-repo container is not upgraded during this upgrade as it is part of the postgres operator release and is updated with the operator.
```
[the remark this answers is not in the document]
Comment (flamingdumpster) [reply hugo/content/Upgrade/_index.md]: Added

**Pull request merged**